Sources/CombineGRPC/Client/StreamingRequestsSubscriber.swift

```diff
@@ -0,0 +1,132 @@
+// Copyright 2019, Vy-Shane Xie
+// Licensed under the Apache License, Version 2.0
+
+import Combine
+import GRPC
+import NIOHPACK
+
+class StreamingRequestsSubscriber<CALL, REQ, RES, DS>: Subscriber, Subscription
+where CALL: StreamingRequestClientCall, CALL.RequestPayload == REQ, CALL.ResponsePayload == RES,
+      DS: Subscriber, DS.Input == RES, DS.Failure == RPCError {
+
+  typealias Input = REQ
+  typealias Failure = Error
+
+  let call: CALL
+  var buffer: DemandBuffer<DS>
+  var subscription: Subscription?
+
+  init(call: CALL, buffer: DemandBuffer<DS>) {
+    self.call = call
+    self.buffer = buffer
+  }
+
+  func receive(subscription: Subscription) {
+    self.subscription = subscription
+
+    // Save any trailingMetadata received before status
+    var trailingMetadata: HPACKHeaders?
+    call.trailingMetadata
+      .whenSuccess { metadata in
+        trailingMetadata = metadata
+      }
+
+    // Send completion as soon as status is available & the subscription has been received
+    call.status
+      .whenSuccess { status in
+
+        switch status.code {
+        case .ok:
+          self.complete()
+
+        default:
+          self.complete(error: RPCError(status: status, trailingMetadata: trailingMetadata))
+        }
+      }
+
+    subscription.request(.max(1))
+  }
+
+  func receive(_ input: REQ) -> Subscribers.Demand {
+
+    call.sendMessage(input)
+      .whenComplete { result in
+
+        switch result {
+        case .success:
+
+          self.subscription?.request(.max(1))
+
+        case .failure(let error):
+
+          self.complete(error: .from(
+            error: error,
+            statusCode: .dataLoss,
+            message: "Bidirectional Request Stream Send End Failed"
+          ))
+
+          self.call.cancel(promise: nil)
```

Comment:
Yes I cancel the client call to effectively end it immediately. That is not at odds with data loss. I nullify the subscriber before cancelling to ensure that any status (regardless of if it is CANCELLED or not) is never reported. This is because I have already discovered and reported an error, the error from the client stream.

You seem to want to honor only GRPC semantics. Remember we are integrating GRPC with Combine. When an error occurs in a Combine publisher you release all outstanding resources and that may include cancelling them.

We _have_ to report the error from the client stream. It is an *error* not a cancellation and it needs to be reported as such.